### DISCUSSION TITLE
Observation about pinwheels and quad edge directionality

### DISCUSSION THREAD

**Discussion by micycle1** [gwlucastrig/Tinfour#80] [Q&A] [answered]
What you're seeing here is a dual-graph of triangulation (overlaid on top of the corresponding triangulation).

On some triangulations, a face from the dual-graph is missing (as seen in the image). When this happens, it's usually a somewhat central face.

![image](https://user-images.githubusercontent.com/9304234/132763517-fca6fd58-0139-455a-91e1-dfa032712dc0.png)

What I think may be happening is that for that single vertex, each of it's connected **base** QuadEdges terminate at it on their B vertex. Since I'm using the pinwheel of each base edge to find which dual-graph edges surround it (to build up the corresponding dual face), this face gets left out (since pinwheels rotate around the A vertex of an edge).

I suspect that for all other vertices, their connected edges terminate at that vertex on a mix of A and B vertices, so my code successfully creates a dual-face for the vertex based on a pinwheel.

![image](https://user-images.githubusercontent.com/9304234/132763824-40a0b00c-2856-4f41-ae5a-18663beca606.png)

So, I wonder whether it was ever a design factor to guarantee that the edges connected to each vertex in a Tin terminate at the vertex using at least one A and one B vertex.

**Comment by gwlucastrig** [comment]
First off, even with a flaw what you're doing is so cool! I am amazed by
the stuff you come up with.

The concept of a base edge really has no mathematical definition that i am
aware of. It's just an artifact of memory management.  I think that, except
at the perimeter,  whether a vertex starts a base edge or a dual edge is
just probability.  So if there are on average six edges per vertex, then
the is a one in 64 chance that the edges starting from a vertex will all be
base edges.

I could be wrong about the true value of the probability because there may
be other factors at work with the edge caching and reuse during insertion.
The process could promote mixing, but does not guarantee it.

Maybe if you tell me your requirements, I can come up with a simple example
code snippet that does what you need.

Gary

P.S. the tin.edges() iterator probably presents all base edges just because
of the way the internal edge pool works. I'll have to check.

**Reply by micycle1** [reply]
It's ok now, but good to know the condition **isn't** assured by the TIN. I've now been able to code around it (if any edge's B vertex hasn't been the A vertex in any pinwheel, then also process that edge's dual).

**Comment by micycle1** [comment, marked as answer]
A workaround was to process edge duals (not just base edges), so the missing pinwheel is picked up as an A vertex.

```
final HashSet<Vertex> seenHubs = new HashSet<>();

edgeDuals.keySet().forEach(edge -> {
	if (hasConstraints && !edge.isConstrainedRegionInterior()) {
		return;
	}

	if (seenHubs.add(edge.getA())) {
		PShape face = dualFace(edge);
		if (face != null) {
			mesh.addChild(face);
		}
	}
	if (seenHubs.add(edge.getB())) {
		PShape face = dualFace(edge.getDual());
		if (face != null) {
			mesh.addChild(face);
		}
	}
});
```

![image](https://user-images.githubusercontent.com/9304234/132908282-af9f6ab8-821d-4244-801b-0c4c3c6d9aca.png)
![image](https://user-images.githubusercontent.com/9304234/132909312-46b9835a-6bfd-4483-ac75-ceec2cf363b6.png)

**Reply by gwlucastrig** [reply]
Wow!  Those images are so cool!

Would be okay if I included one of them on the main Tinfour page?

**Reply by micycle1** [reply]
That's fine! Alternatively you can make your own once I release next version of PGS -- it'll include the necessary things to make the above: mesh-coloring (via graph-coloring), phyllotaxis point distribution and triangulation dual polygon faces.
